ca/provisioner.go

```diff
@@ -0,0 +1,194 @@
+	return p.name
+}
+
+// Kid returns the provisioners key ID.
+func (p *Provisioner) Kid() string {
+	return p.kid
+}
+
+// Token generates a bootstrap token for a subject.
+func (p *Provisioner) Token(subject string) (string, error) {
+	// A random jwt id will be used to identify duplicated tokens
+	jwtID, err := randutil.Hex(64) // 256 bits
+	if err != nil {
+		return "", err
+	}
+
+	notBefore := time.Now()
+	notAfter := notBefore.Add(tokenLifetime)
+	signURL := fmt.Sprintf("%v/1.0/sign", p.caURL)
+
+	tokOptions := []token.Options{
+		token.WithJWTID(jwtID),
+		token.WithKid(p.kid),
+		token.WithIssuer(p.name),
+		token.WithAudience(signURL),
+		token.WithValidity(notBefore, notAfter),
+		token.WithRootCA(p.caRoot),
+		token.WithSANS([]string{subject}),
+	}
+
+	tok, err := provision.New(subject, tokOptions...)
+	if err != nil {
+		return "", err
+	}
+
+	return tok.SignedString(p.jwk.Algorithm, p.jwk.Key)
+}
+
+func decryptProvisionerJWK(encryptedKey string, password []byte) (*jose.JSONWebKey, error) {
+	enc, err := jose.ParseEncrypted(encryptedKey)
+	if err != nil {
+		return nil, err
+	}
+	data, err := enc.Decrypt(password)
+	if err != nil {
+		return nil, err
+	}
+	jwk := new(jose.JSONWebKey)
+	if err := json.Unmarshal(data, jwk); err != nil {
+		return nil, errors.Wrap(err, "error unmarshaling provisioning key")
+	}
+	return jwk, nil
+}
+
+// loadProvisionerJWKByKid retrieves a provisioner key from the CA by key ID and
+// decrypts it using the specified password file.
+func loadProvisionerJWKByKid(kid, caURL, caRoot string, password []byte) (*jose.JSONWebKey, error) {
+	encrypted, err := getProvisionerKey(caURL, caRoot, kid)
+	if err != nil {
+		return nil, err
+	}
+
+	return decryptProvisionerJWK(encrypted, password)
+}
+
+// loadProvisionerJWKByName retrieves the list of provisioners and encrypted key then
+// returns the key of the first provisioner with a matching name that can be successfully
+// decrypted with the specified password file.
+func loadProvisionerJWKByName(name, caURL, caRoot string, password []byte) (key *jose.JSONWebKey, err error) {
+	provisioners, err := getProvisioners(caURL, caRoot)
+	if err != nil {
+		err = errors.Wrap(err, "error getting the provisioners")
+		return
+	}
+
+	for _, provisioner := range provisioners {
+		if provisioner.GetName() == name {
+			if _, encryptedKey, ok := provisioner.GetEncryptedKey(); ok {
+				key, err = decryptProvisionerJWK(encryptedKey, password)
+				if err == nil {
+					return
+				}
+			}
+		}
+	}
+	return nil, errors.Errorf("provisioner '%s' not found (or your password is wrong)", name)
+}
+
+// getRootCAPath returns the path where the root CA is stored based on the
+// STEPPATH environment variable.
+func getRootCAPath() string {
+	return filepath.Join(config.StepPath(), "certs", "root_ca.crt")
+}
+
+// getProvisioners returns the map of provisioners on the given CA.
+func getProvisioners(caURL, rootFile string) (provisioner.List, error) {
```

Comment:
I'm surprised the compiler is smart enough to understand that the function will never continue executing outside of your for loop (you don't need a separate return afterwards). Took me a minute to understand.

Reply:
If there's a return or a break in the for loop, the compiler assumes that you'll get to that at some point, but you can be in an infinite loop if that does not happen.